### ISSUE TITLE
Callable type is reported as not callable when used as a decorator

### ISSUE THREAD

**Issue by sobolevn** [python/mypy#11639]
Example:

```python
from typing import Callable, Union, Any, overload, Protocol

class DCall(Protocol):
    def __call__(self, arg: Union[int, str]) -> None:
        ...

class D:
    def __getattr__(self, attr: str) -> DCall:
        # This method might contain something like during runtime:
        # if attr == '__call__':
        #     return some_func
        # assert False
        ...

def dec_d(f: Callable[..., Any]) -> D:
    return D()

@overload
def f_d(arg: int) -> None: ...
@overload
def f_d(arg: str) -> None: ...
@dec_d
def f_d(arg: Union[int, str]) -> None: ...

```

Outputs:

```
out/ex.py:18: error: "D" not callable
```

This happens here: 

https://github.com/python/mypy/blob/4f59ca490072a45dce5839e6397791d79dcf3352/mypy/checker.py#L510-L520

Refs https://github.com/python/mypy/pull/11638
Refs https://github.com/python/mypy/pull/11630

**Comment by sobolevn** [comment]
Mypy is right here. See https://github.com/python/mypy/pull/11638#pullrequestreview-817603627
